### ISSUE TITLE
security component is not initialized by the time user components are started

### ISSUE THREAD

**Issue by cordellcalitz** [happner/happner#60]
When calling a security method inside a component start method, the following error is returned:
"Error: security module not initialized, is your datalayer configured to be secure?"

in the config the data layer is set up with secure=true

Calling the security methods after start (in another method or with a setTimeout) works.

**Comment by southbite** [comment]
mmmm.... not good, I think Richard may have raised this a while back - I'l
take a look...

On Tue, Dec 15, 2015 at 11:51 AM, cordellcalitz <notifications@github.com>
wrote:

> When calling a security method inside a component start method, the
> following error is returned:
> "Error: security module not initialized, is your datalayer configured to
> be secure?"
>
> in the config the data layer is set up with secure=true
>
> Calling the security methods after start (in another method or with a
> setTimeout) works.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/happner/happner/issues/60>.
>



-- 

SIMON BISHOP

DIRECTOR

+27 (0) 83 282 0703

|     simon@tenacious.digital <brianna@tenacious.digital>

twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>

|     skype: sjbishop

201 Salt Circle Arcade, Salt River, Cape Town, 7925

www.tenacious.digital

**Comment by southbite** [comment]
Mmmm- may take some work, 2 solutions to investigate:

(1) the component does its initialisation as part of its construction
(2) we have the notion of component start order - so we can attach
priorities to components - at the moment the _eachComponentDo method runs
the start method on the components - what we want to do is pass a
collection of priorities that match component names - so that they can be
operated on in order.

anyones thoughts here are welcome...

On Tue, Dec 15, 2015 at 12:09 PM, Simon Bishop <simon@tenacious.digital>
wrote:

> mmmm.... not good, I think Richard may have raised this a while back - I'l
> take a look...
>
> On Tue, Dec 15, 2015 at 11:51 AM, cordellcalitz <notifications@github.com>
> wrote:
>
>> When calling a security method inside a component start method, the
>> following error is returned:
>> "Error: security module not initialized, is your datalayer configured to
>> be secure?"
>>
>> in the config the data layer is set up with secure=true
>>
>> Calling the security methods after start (in another method or with a
>> setTimeout) works.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/happner/happner/issues/60>.
>>
>
>
>
> --
>
> SIMON BISHOP
>
> DIRECTOR
>
> +27 (0) 83 282 0703
>
> |     simon@tenacious.digital <brianna@tenacious.digital>
>
> twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>
>
> |     skype: sjbishop
>
> 201 Salt Circle Arcade, Salt River, Cape Town, 7925
>
> www.tenacious.digital
>



-- 

SIMON BISHOP

DIRECTOR

+27 (0) 83 282 0703

|     simon@tenacious.digital <brianna@tenacious.digital>

twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>

|     skype: sjbishop

201 Salt Circle Arcade, Salt River, Cape Town, 7925

www.tenacious.digital

**Comment by nomilous** [comment]
Javascript preserves key order.

And the order appears to originate here:

https://github.com/happner/happner/blob/master/lib/mesh.js#L765

Object.key( config.components ) produces an array of keys in the order that
they were created / declared.

So instead of adding more config we can rather let the declared order of
components determine the order.

*The problem is that attachSystemComponents() is adding the system
components to an existing list - hence to the end of the list they go.*

The correction to create a new reorderedComponents = {} in the
attachSystemComponents(), put the system components onto it first, then add
the others in the same order and assign config.components =
reorderedComponents

Unless there is a better way to put keys into the front on an existing
object (i doubt it)

**Comment by nomilous** [comment]
I think the thing I raised a while back is the network is ready and listening before the components have started: https://github.com/happner/happner/issues/46

**Comment by cordellcalitz** [comment]
I'm not familiar with the internals, but could we not first process all the system components and only process the user/config declared components once that is done? So basically keep two lists and only process the second list after the first. Rather than adding them all to one list and hoping the order is correct.

**Comment by southbite** [comment]
Well - it only happns once - so cost isn't a factor. That being said, being
able to configure the order you want your components started is actually a
nice feature - what I propose is that we default the order config to start
the system components first - this means a modification to the _eachComponentDo
method to take an object that matches keys to priorities - we then have a
method that returns a reordered list - which is what get iterated through -
system components can be prioritised with negative numbers (items in the
configuration file must be configured with 0 and above)

Do you think I am adding unnecessary complexity here? Its not a lot of
code, and it wont add any bloat to the config if we default the system
components... but it will come in handy when we load dependant components
that need certain sequences of predecessors started before they can be
started?




On Tue, Dec 15, 2015 at 10:14 PM, Richard Emslie <notifications@github.com>
wrote:

> Javascript preserves key order.
>
> And the order appears to originate here:
>
> https://github.com/happner/happner/blob/master/lib/mesh.js#L765
>
> Object.key( config.components ) produces an array of keys in the order that
> they were created / declared.
>
> So instead of adding more config we can rather let the declared order of
> components determine the order.
>
> *The problem is that attachSystemComponents() is adding the system
> components to an existing list - hence to the end of the list they go.*
>
> The correction to create a new reorderedComponents = {} in the
> attachSystemComponents(), put the system components onto it first, then add
> the others in the same order and assign config.components =
> reorderedComponents
>
> Unless there is a better way to put keys into the front on an existing
> object (i doubt it)
>
>
>
> On Tue, Dec 15, 2015 at 9:35 PM, southbite <notifications@github.com>
> wrote:
>
> > Mmmm- may take some work, 2 solutions to investigate:
> >
> > (1) the component does its initialisation as part of its construction
> > (2) we have the notion of component start order - so we can attach
> > priorities to components - at the moment the _eachComponentDo method runs
> > the start method on the components - what we want to do is pass a
> > collection of priorities that match component names - so that they can be
> > operated on in order.
> >
> > anyones thoughts here are welcome...
> >
> > On Tue, Dec 15, 2015 at 12:09 PM, Simon Bishop <simon@tenacious.digital>
> > wrote:
> >
> > > mmmm.... not good, I think Richard may have raised this a while back -
> > I'l
> > > take a look...
> > >
> > > On Tue, Dec 15, 2015 at 11:51 AM, cordellcalitz <
> > notifications@github.com>
> > > wrote:
> > >
> > >> When calling a security method inside a component start method, the
> > >> following error is returned:
> > >> "Error: security module not initialized, is your datalayer configured
> to
> > >> be secure?"
> > >>
> > >> in the config the data layer is set up with secure=true
> > >>
> > >> Calling the security methods after start (in another method or with a
> > >> setTimeout) works.
> > >>
> > >> —
> > >> Reply to this email directly or view it on GitHub
> > >> <https://github.com/happner/happner/issues/60>.
> > >>
> > >
> > >
> > >
> > > --
> > >
> > > SIMON BISHOP
> > >
> > > DIRECTOR
> > >
> > > +27 (0) 83 282 0703
> > >
> > > | simon@tenacious.digital <brianna@tenacious.digital>
> > >
> > > twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>
> > >
> > > | skype: sjbishop
> > >
> > > 201 Salt Circle Arcade, Salt River, Cape Town, 7925
> > >
> > > www.tenacious.digital
> > >
> >
> >
> >
> > --
> >
> > SIMON BISHOP
> >
> > DIRECTOR
> >
> > +27 (0) 83 282 0703
> >
> > | simon@tenacious.digital <brianna@tenacious.digital>
> >
> > twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>
> >
> > | skype: sjbishop
> >
> > 201 Salt Circle Arcade, Salt River, Cape Town, 7925
> >
> > www.tenacious.digital
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/happner/happner/issues/60#issuecomment-164868472>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/happner/happner/issues/60#issuecomment-164883215>.
>



-- 

SIMON BISHOP

DIRECTOR

+27 (0) 83 282 0703

|     simon@tenacious.digital <brianna@tenacious.digital>

twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>

|     skype: sjbishop

201 Salt Circle Arcade, Salt River, Cape Town, 7925

www.tenacious.digital

**Comment by southbite** [comment]
mmm - yeah, a bunch of related things here. Cordell - could you put
together a unit test that demonstrates this issue, I'll look at solving it
as high in the call stack as possible, it may be an adjustment of happn
that needs to be made.

On Tue, Dec 15, 2015 at 10:20 PM, Richard Emslie <notifications@github.com>
wrote:

> I think the thing I raised a while back is the network is ready and
> listening before the components have started: #46
> <https://github.com/happner/happner/issues/46>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/happner/happner/issues/60#issuecomment-164885273>.
>



-- 

SIMON BISHOP

DIRECTOR

+27 (0) 83 282 0703

|     simon@tenacious.digital <brianna@tenacious.digital>

twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>

|     skype: sjbishop

201 Salt Circle Arcade, Salt River, Cape Town, 7925

www.tenacious.digital

**Comment by JohanvdWest** [comment]
I agree that it will be a nice feature to set the start order of user components, but it must be optional.

We need to make sure that the system components start up before any user components and are ready to use.

**Comment by cordellcalitz** [comment]
Unit test created on branch: bug/#60
run: mocha test/bug_60.js

**Comment by southbite** [comment]
Hi guys,

I fixed this in a more simple manner for now, I think Richard will be more
happy with this :) - I just modified how the system components are attached
to the config, so now I create a new component list - attach the system
components first, then iterate through the old config and attach the
components on it to the new config.

What this means is that components with start methods always start in the
order that they are listed in the config, but always after the key system
components - which are added to the config first.

I checked it into the test branch - the test now passes Cordell - let me
know if all is good with this solution - and I will merge into master.

Cheerio!


On Thu, Dec 17, 2015 at 11:56 AM, cordellcalitz <notifications@github.com>
wrote:

> Unit test created on branch: bug/#60
> <https://github.com/happner/happner/issues/60>
> run: mocha test/bug_60.js
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/happner/happner/issues/60#issuecomment-165404258>.
>



-- 

SIMON BISHOP

DIRECTOR

+27 (0) 83 282 0703

|     simon@tenacious.digital <brianna@tenacious.digital>

twitter: @_tenacious_ <http://www.twitter.com/_tenacious_>

|     skype: sjbishop

201 Salt Circle Arcade, Salt River, Cape Town, 7925

www.tenacious.digital

**Comment by cordellcalitz** [comment]
Hi Simon,
The fix seems to be working, thanks.

**Comment by southbite** [comment]
Fixed release 1.2.7
